src/btree/bt_prefetch.c

```diff
@@ -0,0 +1,94 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __wt_btree_prefetch --
+ *     Pre-load a set of pages into the cache. This session holds a hazard pointer on the ref passed
+ *     in, so there must be a valid page and a valid parent page (though that parent could change if
+ *     a split happens).
+ */
+int
+__wt_btree_prefetch(WT_SESSION_IMPL *session, WT_REF *ref)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    WT_REF *next_ref;
+    uint64_t block_preload;
+
+    conn = S2C(session);
+    block_preload = 0;
+
+    WT_ASSERT_ALWAYS(session, F_ISSET(ref, WT_REF_FLAG_LEAF),
+      "Pre-fetch starts with a leaf page and reviews the parent");
+
+    WT_ASSERT_ALWAYS(session, __wt_session_gen(session, WT_GEN_SPLIT) != 0,
+      "Pre-fetch requires a split generation to traverse internal page(s)");
+
+    session->prefetch_prev_ref = ref;
+    /* Load and decompress a set of pages into the block cache. */
+    WT_INTL_FOREACH_BEGIN (session, ref->home, next_ref) {
+        /* Don't let the pre-fetch queue get overwhelmed. */
+        if (conn->prefetch_queue_count > WT_MAX_PREFETCH_QUEUE ||
+          block_preload > WT_PREFETCH_QUEUE_PER_TRIGGER)
+            break;
+
+        /*
+         * Skip queuing pages that are already in cache or are internal. They aren't the pages we
+         * are looking for. This pretty much assumes that all children of an internal page remain in
+         * cache during the scan. If a previous pre-fetch of this internal page read a page in, then
+         * that page was evicted and now a future page wants to be pre-fetched, this algorithm needs
+         * a tweak. It would need to remember which child was last queued and start again from
+         * there, rather than this approximation which assumes recently pre-fetched pages are still
+         * in cache.
+         */
+        if (next_ref->state == WT_REF_DISK && F_ISSET(next_ref, WT_REF_FLAG_LEAF)) {
+            ret = __wt_conn_prefetch_queue_push(session, next_ref);
+            if (ret == 0)
+                ++block_preload;
+            else if (ret != EBUSY)
+                WT_RET(ret);
+        }
+    }
+    WT_INTL_FOREACH_END;
+
+    WT_STAT_CONN_INCRV(session, block_prefetch_pages_queued, block_preload);
+    return (0);
+}
+
+/*
+ * __wt_prefetch_page_in --
+ *     Does the heavy lifting of reading a page into the cache. Immediately releases the page since
+ *     reading it in is the useful side effect here. Must be called while holding a dhandle.
+ */
+int
+__wt_prefetch_page_in(WT_SESSION_IMPL *session, WT_PREFETCH *pf)
+{
+    WT_ADDR_COPY addr;
+
+    WT_ASSERT_ALWAYS(
+      session, pf->ref->home == pf->first_home, "The home changed while queued for pre-fetch");
```

Comment:
Yes it is possible for a split to happen here, but I think in that case the `ref` will become invalid. I think @agorrod  might remember the details better of why we can't do prefetching here.

Reply:
Right, but the problem is `home` changing, so maybe the parent got split. I don't mind if we bail, but making it an assert feels a bit strong since it seems like a legitimate thing that could happen.

Reply:
I'll turn this into a verbose message for now.